sys/net/link_layer/ieee802154/security.c

```diff
@@ -0,0 +1,503 @@
+    uint8_t aux_size = _get_aux_hdr_size(security_level, key_mode);
+    uint8_t mac_size = _mac_size(security_level);
+    /* remember that the frame counter was stored in little endian */
+    uint32_t frame_counter = byteorder_ntohl((be_uint32_t)
+                                byteorder_ltobl((le_uint32_t)aux->fc));
+
+    if (security_level == IEEE802154_SCF_SECLEVEL_NONE) {
+        *payload = header + *header_size;
+        *payload_size = frame_size - *header_size;
+        *mic = NULL;
+        *mic_size = 0;
+        return IEEE802154_SEC_OK;
+    }
+
+    *payload_size = frame_size - *header_size - aux_size - mac_size;
+    *payload = header + *header_size + aux_size;
+    *mic_size = mac_size;
+    *mic = header + frame_size - mac_size;
+
+    /* attempt to find the decryption key */
+    const uint8_t *key;
+    if (!(key = _get_decryption_key(ctx, header, *header_size, aux))) {
+        return -IEEE802154_SEC_NO_KEY;
+    }
+    _set_key(ctx, key);
+
+    /* better decrypt block by block, than holding a large buffer because I had stack issues */
+    uint8_t tmp1[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t tmp2[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    const uint8_t *a = header;
+    uint8_t *c = *payload;
+    uint16_t a_len = *header_size + aux_size;
+    uint16_t c_len = *payload_size;
+    uint8_t *mac = *mic;
+    uint8_t off;
+    ieee802154_ccm_block_t ccm; /* Ai or Bi */
+    _init_ctr_A0(&ccm, frame_counter, security_level, src_address);
+
+    /* TODO:
+       A better implementation would check if the received frame counter is
+       greater then the frame counter that has previously been received from
+       the other endpoint. This is done to protect against replay attacks.
+       But we do not store this information because we also do not have
+       a proper key store, to avoid complexity on embedded devices. */
+
+    /* decrypt MIC */
+    if (mac_size) {
+        _ecb(ctx, tmp1, tmp2, mac, (uint8_t *)&ccm, mac_size);
+    }
+    /* decrypt cipher */
+    if (_req_encryption(security_level)) {
+        for (off = 0; off < c_len;) {
+            _advance_ctr_Ai(&ccm);
+            off += _ecb(ctx, tmp1, tmp2, &c[off], (uint8_t *)&ccm, c_len - off);
+        }
+    }
+    /* check MIC */
+    if (_req_mac(security_level)) {
+        _init_cbc_B0(&ccm, frame_counter, security_level, c_len, mac_size, src_address);
+        memset(tmp2, 0, sizeof(tmp2));
+        _cbc_next(ctx, tmp2, tmp1, (uint8_t *)&ccm, sizeof(ccm));
+        byteorder_htobebufs(tmp1, a_len);
+        off = _min(sizeof(tmp1) - sizeof(uint16_t), a_len);
+        memcpy(tmp1 + sizeof(uint16_t), a, off);
+        _cbc_next(ctx, tmp2, tmp1, tmp1, sizeof(uint16_t) + off);
+        for (;off < a_len;) {
+            off += _cbc_next(ctx, tmp2, tmp1, &a[off], a_len - off);
+        }
+        for (off = 0; off < c_len;) {
+            off += _cbc_next(ctx, tmp2, tmp1, &c[off], c_len - off);
+        }
```

Comment:
Is there some opportunity to share code with the encrypt path? 
It looks like the MIC calculation could be done in a shared helper function.

Reply:
Yes I could probably make it another function `_comp_mic()` or something like that.